File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        for(int num : nums){ map.put(num, map.getOrDefault(num, 0) + 1); }
        
        Queue<Integer> heap = new PriorityQueue<>((a, b) -> map.get(b) - map.get(a));
        for(int key : map.keySet()){ heap.add(key); }
        
        List<Integer> list = new ArrayList<>();
        for(int i = 0; i < k; i++){
            list.add(heap.poll());
        }
        
        int[] result = new int[list.size()];        
        for(int i = 0; i < list.size() ; i++){
            result[i] =list.get(i);            
        }        

        return result;
    }
}

File: 347-top-k-frequent-elements/top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
    if(nums.length === 1){
        return nums
    }
    
    const map = new Map();
    const result = [];

    for(let i = 0; i < nums.length; i ++){
        map.set(nums[i], (map.get(nums[i]) || 0) + 1)
    }

    const count = new Array(nums.length);

    for (let [key, value] of map) {
        if(!count[value]){
            count[value] = new Array();
        }
        count[value].push(key);
    }

    let right = count.length;

    while(result.length < k){
        if(!count[right]){
            right--;
            continue;
        } else{
            if(count[right].length === 1){
                result.push(count[right][0])
            } else{
                for(let i = 0; i < count[right].length; i++){
                    if(result.length === k){
                        break;
                    }
                    result.push(count[right][i])
                }
            }

            right--;
        }

    }

    return result;

};